### PR TITLE
Lp 557 adding delegates to permissions

### PR DESCRIPTION
**Pull request by anujshah108**
fixing function name/modifier and moving to branch with jira name

### CODE REVIEW

File: contracts/LinniaPermissions.sol [hunk 1]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
         return permissions[dataHash][viewer].canAccess;
     }
 
+    /// Add a delegate for a user's permissions
+    /// @param delegate the address of the delegate being added by user
+    function addDelegate(address delegate)
+    onlyUser
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: It may be best to make this onlyOwner. So only the admin can add delegates.

File: contracts/LinniaPermissions.sol [hunk 2]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
     /// @param viewer the user being permissioned to view the data
     /// @param dataUri the ipfs path of the re-encrypted data
-    function grantAccess(bytes32 dataHash, address viewer, string dataUri)
-    onlyUser
-    onlyRecordOwnerOf(dataHash)
+    function grantAccess(
+        bytes32 dataHash, address viewer, string dataUri)
+        onlyUser
+        onlyRecordOwnerOf(dataHash, msg.sender)
+        public
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: Consider making external

File: contracts/LinniaPermissions.sol [hunk 3]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+    /// Called by delegate to the owner of the record.
+    /// @param dataHash the data hash of the linnia record
+    /// @param viewer the user being permissioned to view the data
+    /// @param owner the owner of the linnia record
+    /// @param dataUri the ipfs path of the re-encrypted data
+    function grantAccessbyDelegate(
+        bytes32 dataHash, address viewer, address owner, string dataUri)
+        onlyDelegate(owner)
+        onlyRecordOwnerOf(dataHash, owner)
+        public
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: consider making `external`

File: contracts/LinniaPermissions.sol [hunk 4]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
-    onlyRecordOwnerOf(dataHash)
+    function grantAccess(
+        bytes32 dataHash, address viewer, string dataUri)
+        onlyUser
+        onlyRecordOwnerOf(dataHash, msg.sender)
+        public
+        returns (bool)
+    {
+        require(
+            _grantAccess(dataHash, viewer, msg.sender, dataUri) == true
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: is ` == true` necessary?

File: contracts/LinniaPermissions.sol [hunk 5]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+    /// @param dataUri the ipfs path of the re-encrypted data
+    function grantAccessbyDelegate(
+        bytes32 dataHash, address viewer, address owner, string dataUri)
+        onlyDelegate(owner)
+        onlyRecordOwnerOf(dataHash, owner)
+        public
+        returns (bool)
+    {
+        require(
+            _grantAccess(dataHash, viewer, owner, dataUri) == true
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: is ` == true` necessary?

File: contracts/LinniaPermissions.sol [hunk 6]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+        );
+        return true;
+    }
+
+    /// Revoke a viewer access to a linnia record
+    /// Note that this does not necessarily remove the file from storage
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: consider using
`...  the **data** from storage`

File: contracts/LinniaPermissions.sol [hunk 7]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+    /// Called by owner of the record.
+    /// @param dataHash the data hash of the linnia record
+    /// @param viewer the user that has permission to view the data
+    function revokeAccess(
+        bytes32 dataHash, address viewer)
+        onlyUser
+        onlyRecordOwnerOf(dataHash, msg.sender)
+        public
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: consider making `external`

File: contracts/LinniaPermissions.sol [hunk 8]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+    /// Note that this does not necessarily remove the file from storage
+    /// Called by delegate to the owner of the record.
+    /// @param dataHash the data hash of the linnia record
+    /// @param viewer the user that has permission to view the data
+    /// @param owner the owner of the linnia record
+    function revokeAccessbyDelegate(
+        bytes32 dataHash, address viewer, address owner)
+        onlyDelegate(owner)
+        onlyRecordOwnerOf(dataHash, owner)
+        public
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: consider making `external`

File: contracts/LinniaPermissions.sol [hunk 9]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+        returns (bool)
+    {
+        require(
+            _revokeAccess(dataHash, viewer, owner) == true
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: is == true necessary?

File: contracts/LinniaPermissions.sol [hunk 10]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+        );
+        return true;
+    }
+
+    /// Internal function to give a viewer access to a linnia record
+    /// Called by external functions
+    /// @param dataHash the data hash of the linnia record
+    /// @param viewer the user being permissioned to view the data
+    /// @param dataUri the ipfs path of the re-encrypted data
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: remove `ipfs`

File: contracts/LinniaPermissions.sol [hunk 11]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+    /// @param owner the owner of the linnia record
+    /// @param dataUri the ipfs path of the re-encrypted data
+    function grantAccessbyDelegate(
+        bytes32 dataHash, address viewer, address owner, string dataUri)
+        onlyDelegate(owner)
+        onlyRecordOwnerOf(dataHash, owner)
+        public
+        returns (bool)
+    {
+        require(
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: consider adding :
`require(owner != address(0))`

File: contracts/LinniaPermissions.sol [hunk 12]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
 
+    /// Add a delegate for a user's permissions
+    /// @param delegate the address of the delegate being added by user
+    function addDelegate(address delegate)
+    onlyUser
+    whenNotPaused
+    external
+    returns (bool)
+    {
+        delegates[msg.sender][delegate] = true;
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: Add the following:
require(delegate != address(0));
require(delegate != msg.sender);
require(delegate != this);

File: contracts/LinniaPermissions.sol [hunk 13]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+    /// @param viewer the user that has permission to view the data
+    function revokeAccess(
+        bytes32 dataHash, address viewer)
+        onlyUser
+        onlyRecordOwnerOf(dataHash, msg.sender)
+        public
+        returns (bool)
+    {
+        require(
+            _revokeAccess(dataHash, viewer, msg.sender) == true
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: is ` == true` necessary?

File: package.json [hunk 14]
```diff
@@ -39,18 +39,19 @@
     "babel-preset-stage-3": "^6.24.1",
     "babel-register": "^6.26.0",
     "eslint": "^5.0.1",
-    "eslint-config-airbnb": "^17.0.0",
+    "eslint-config-airbnb": "^17.1.0",
```
Comment (godfreyhobbs) [inline package.json]: Can you revert these changes?  How they are related to contract changes?

File: contracts/LinniaPermissions.sol [hunk 15]
```diff
@@ -16,25 +16,33 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
     }
 
-    modifier onlyRecordOwnerOf(bytes32 dataHash) {
-        require(hub.recordsContract().recordOwnerOf(dataHash) == msg.sender);
+    modifier onlyRecordOwnerOf(bytes32 dataHash, address owner) {
+        require(hub.recordsContract().recordOwnerOf(dataHash) == owner);
+        _;
+    }
+
+    modifier onlyDelegate(address owner) {
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: consider renaming `onlyWhenSenderIsDelegate` or something that relates to msg.sender.

File: contracts/LinniaPermissions.sol [hunk 16]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+    /// @param dataHash the data hash of the linnia record
+    /// @param viewer the user that has permission to view the data
+    /// @param owner the owner of the linnia record
+    function revokeAccessbyDelegate(
+        bytes32 dataHash, address viewer, address owner)
+        onlyDelegate(owner)
+        onlyRecordOwnerOf(dataHash, owner)
+        public
+        returns (bool)
+    {
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: consider adding :
`require(owner != address(0))`

File: contracts/LinniaPermissions.sol [hunk 17]
```diff
@@ -59,16 +68,101 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+        );
+        return true;
+    }
+
+    /// Give a viewer access to a linnia record
+    /// Called by delegate to the owner of the record.
+    /// @param dataHash the data hash of the linnia record
+    /// @param viewer the user being permissioned to view the data
+    /// @param owner the owner of the linnia record
+    /// @param dataUri the ipfs path of the re-encrypted data
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: please remove term `ipfs`

File: test/4_permissions_test.js [hunk 18]
```diff
@@ -102,6 +118,38 @@ contract('LinniaPermissions', accounts => {
...
+          from: delegate
+        })
+      );
+    });
+    it('should allow delegate to grant access', async () => {
+      const fakeIpfsHash = eutil.bufferToHex(crypto.randomBytes(32));
+      await instance.addDelegate(delegate, {from: user1});
+      const tx = await instance.grantAccessbyDelegate(
+        testDataHash1,
+        provider2,
```
Comment (godfreyhobbs) [inline test/4_permissions_test.js]: It is a bit confusing why you are granting access to a `provider`.  Consider renaming `viewer`.

File: test/4_permissions_test.js [hunk 19]
```diff
@@ -102,6 +118,38 @@ contract('LinniaPermissions', accounts => {
...
     });
+    it('should not allow non-delegate to grant access', async () => {
+      const fakeIpfsHash = eutil.bufferToHex(crypto.randomBytes(32));
+      await assertRevert(
+        instance.grantAccessbyDelegate(testDataHash1, provider2, user1, fakeIpfsHash, {
+          from: delegate
+        })
+      );
+      await assertRevert(
+        instance.grantAccessbyDelegate(testDataHash2, provider1, provider2, fakeIpfsHash, {
```
Comment (godfreyhobbs) [inline test/4_permissions_test.js]: Not clear why you are testing here.  Maybe rename `delegate` to `nondelegate`

File: test/4_permissions_test.js [hunk 20]
```diff
@@ -157,6 +205,31 @@ contract('LinniaPermissions', accounts => {
...
+      assert.equal(tx.logs[0].args.owner, user1);
+      assert.equal(tx.logs[0].args.viewer, provider2);
+      assert.equal(tx.logs[0].args.sender, delegate);
+      const perm = await instance.permissions(testDataHash1, provider2);
+      assert.equal(perm[0], false);
+      assert.equal(perm[1], '');
+    });
+    it('should not allow non-delegate to revoke access to data', async () => {
+      await assertRevert(
+        instance.revokeAccessbyDelegate(testDataHash1, provider2, user1, { from: delegate })
```
Comment (godfreyhobbs) [inline test/4_permissions_test.js]: not clear what this is testing.

File: contracts/LinniaPermissions.sol [hunk 21]
```diff
@@ -16,25 +16,33 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
         string dataUri;
     }
 
-    event LinniaAccessGranted(bytes32 indexed dataHash, address indexed owner,
-        address indexed viewer
+    event LinniaAccessGranted(bytes32 dataHash, address indexed owner,
```
[the remark this answers is not in the document]
Comment (anujshah108) [reply contracts/LinniaPermissions.sol]: events only allow three indexed params

File: contracts/LinniaPermissions.sol [hunk 22]
```diff
@@ -16,25 +16,33 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+        address indexed viewer, address indexed sender
     );
-    event LinniaAccessRevoked(bytes32 indexed dataHash, address indexed owner,
```
[the remark this answers is not in the document]
Comment (anujshah108) [reply contracts/LinniaPermissions.sol]: same as above

File: contracts/LinniaPermissions.sol (hunk 1 above)
Comment (anujshah108) [reply contracts/LinniaPermissions.sol]: I think a user should be able to add their own delegates and also has the authority to only allow delegates they approve.

File: contracts/LinniaPermissions.sol (hunk 4 above)
Comment (anujshah108) [reply contracts/LinniaPermissions.sol]: nice catch!

File: package.json (hunk 14 above)
Comment (anujshah108) [reply package.json]: why not include the newest versions of the library? they are working well

File: contracts/LinniaPermissions.sol (hunk 21 above)
[the remark this answers is not in the document]
Comment (anujshah108) [reply contracts/LinniaPermissions.sol]: I will include dataHash and remove index of sender. That is probably the least important.

File: test/4_permissions_test.js [hunk 23]
```diff
@@ -16,7 +16,7 @@ const testDataUri1 = 'QmUMqi1rr4Ad1eZ3ctsRUEmqK2U3CyZqpetUe51LB9GiAM';
 const testDataUri2 = 'QmUoCHEZqSuYhr9fV1c2b4gLASG2hPpC2moQXQ6qzy697d';
 const testMetadata = 'KEYWORDS';
 
-contract('LinniaPermissions', accounts => {
+contract.only('LinniaPermissions', accounts => {
```
Comment (godfreyhobbs) [inline test/4_permissions_test.js]: do not check this in

File: contracts/LinniaPermissions.sol (hunk 1 above)
Comment (MaxBlaushild) [reply contracts/LinniaPermissions.sol]: I second this. The user needs to be able to set the ddex as a delegate for the registration flow to work

Comment (godfreyhobbs) [reply contracts/LinniaPermissions.sol]: Ok. Let go with that.  Let try to make it easy for end users to see the list of their delegates.

File: contracts/LinniaPermissions.sol [hunk 24]
```diff
@@ -59,19 +68,105 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
...
+        emit LinniaPermissionDelegateAdded(msg.sender, delegate);
+        return true;
+    }
+    
+    /// Give a viewer access to a linnia record
+    /// Called by owner of the record.
+    /// @param dataHash the data hash of the linnia record
+    /// @param viewer the user being permissioned to view the data
+    /// @param dataUri the path of the re-encrypted data
+    function grantAccess(
```
Comment (godfreyhobbs) [inline contracts/LinniaPermissions.sol]: formatting looks off.

File: package.json (hunk 14 above)
Comment (godfreyhobbs) [reply package.json]: ok

File: package.json [hunk 25]
```diff
@@ -39,18 +39,19 @@
...
+    "eslint-plugin-jsx-a11y": "^6.1.1",
+    "eslint-plugin-react": "^7.11.0",
     "ethereumjs-util": "5.2.0",
     "ganache-cli": "^6.0.3",
     "multihashes": "^0.4.13",
     "solidity-coverage": "0.5.2",
     "solium": "1.1.7",
-    "truffle": "4.1.11"
+    "truffle": "4.1.11",
+    "webpack": "^4.20.2"
```
Comment (godfreyhobbs) [inline package.json]: Can you remove `webpack`

File: test/4_permissions_test.js [hunk 26]
```diff
@@ -67,6 +69,20 @@ contract('LinniaPermissions', accounts => {
...
   });
+  describe('delegate', () => {
+    it('should not let non-user set a delegate', async () => {
+      await assertRevert(
+        instance.addDelegate(delegate, {from: nonUser})
+      );
+    });
+    it('should set a delegate', async () => {
+      const tx = await instance.addDelegate(delegate, {from: user1});
+      assert.equal(await instance.delegates.call(user1, delegate), true);
```
Comment (godfreyhobbs) [inline test/4_permissions_test.js]: switch the order of these lines.

Comment (anujshah108) [reply test/4_permissions_test.js]: can you explain your reasoning behind this please? we set the delegate in 79 and then test it in 80.

Comment (godfreyhobbs) [reply test/4_permissions_test.js]: can you move line 80 to line 84

Comment (anujshah108) [reply test/4_permissions_test.js]: cool
